### PR TITLE
Apply classes

### PR DESCRIPTION
**Pull request by jacobobryant**
I've broken this into two steps: (1) map over the relevant rules, replacing instances of target classes with the alias class; (2) merge the new rules where possible to reduce output file size.

1 is done with postwalk rather than by traversing garden's grammar. So the implementation is super simple and should work in all cases, assuming you don't reuse css keywords as class names. Since 2 is just for optimization, it doesn't have to be comprehensive. I've made it understand enough garden grammar to cover the test case, which I believe should also cover everything else produced by girouette.

### CODE REVIEW

**Comment by green-coder** [comment]
Thank you for the PR. I will review it during the weekend.

Related to #35

File: lib/girouette/src/girouette/garden/util.cljc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+(defn apply-class-rules [garden alias-class classes]
+  (merge-rules
+    (keep (fn [rule]
+            (let [new-rule (walk/postwalk (fn [x]
+                                            (if (contains? classes x)
+                                              alias-class
+                                              x))
+                                          rule)]
+              (when (not= rule new-rule)
```
Comment (green-coder) [inline lib/girouette/src/girouette/garden/util.cljc:51]: In which situation would `(= rule new-rule)` ?

File: lib/girouette/test/girouette/garden/util_test.cljc [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            [girouette.tw.default-api :refer [class-name->garden]]
+            [girouette.tw.common :refer [dot]]
+            [girouette.garden.util :refer [apply-class-rules]]))
+
+(deftest apply-class-rules-test
+  (are [class-name classes expected-garden]
+       (= expected-garden
+          (apply-class-rules (mapv class-name->garden classes)
+                                   (dot class-name)
+                                   (set (map dot classes))))
```
Comment (green-coder) [inline lib/girouette/test/girouette/garden/util_test.cljc:12]: Small indentation problem here - I will fix it, no need to update to code.

File: lib/processor/src/girouette/processor.clj [hunk 3]
```diff
@@ -157,7 +167,15 @@
...
-            all-garden-defs (into predef-garden (keep (:garden-fn @config)) (sort all-css-classes))]
+            all-garden-defs (into predef-garden (keep (:garden-fn @config)) (sort all-css-classes))
+            all-garden-defs (reduce (fn [garden [cls classes]]
+                                      (into garden
+                                            (util/apply-class-rules
+                                              garden
+                                              (first (normalize-classes cls))
+                                              (set (mapcat normalize-classes classes)))))
+                                    all-garden-defs
+                                    @(requiring-resolve apply-classes))]
```
Comment (green-coder) [inline lib/processor/src/girouette/processor.clj:178]: When possible, I prefer to use alternatives to reduce. I will change that bit.

File: lib/girouette/src/girouette/garden/util.cljc (hunk 1 above)
Comment (jacobobryant) [reply lib/girouette/src/girouette/garden/util.cljc:51]: It'll be true for rules that aren't included by the applied class. e.g. if you have `"btn"` => `"p-2" "rounded" "bg-blue-500"` and you have `"mx-2"` elsewhere in your css, then `(= rule new-rule)` is true for the `"mx-2"` rule.

File: lib/girouette/test/girouette/garden/util_test.cljc (hunk 2 above)
Comment (jacobobryant) [reply lib/girouette/test/girouette/garden/util_test.cljc:12]: Ah, looks like my brain tricked me on this one :)

File: lib/processor/src/girouette/processor.clj (hunk 3 above)
Comment (green-coder) [inline lib/processor/src/girouette/processor.clj:178]: Could you please give an example of how the apply-classes parameter is used?
What kind of values is returned by `@(requiring-resolve apply-classes)`?

Comment (jacobobryant) [reply lib/processor/src/girouette/processor.clj:178]: it is a map from class -> classes, e.g. `{"btn" ["p-2" "rounded"]}`

File: lib/girouette/src/girouette/garden/util.cljc [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+       (util/group-by :ident :value)
+       (map (fn [[ident values]]
+              (merge-similar-rules ident values)))))
+
+(defn apply-class-rules
+  "Returns a collection of garden rules defining `target-class-name` as an aggregation of `gi-garden-rules`.
+   `target-class-name` is the dotted CSS class name which we want to define.
+   `gi-garden-rules` is an ordered collection of garden rules generated by Girouette.
+   `gi-class-names` is an ordered collection of the CSS dotted class-names defined in `gi-garden-rules`."
+  [target-class-name gi-garden-rules gi-class-names]
```
Comment (green-coder) [inline lib/girouette/src/girouette/garden/util.cljc:47]: I exchanged the position of the first 2 arguments, not sure of to justify it .. sorry

File: lib/girouette/src/girouette/garden/util.cljc (hunk 1 above)
Comment (green-coder) [reply lib/girouette/src/girouette/garden/util.cljc]: You are right, I will restore that test.

File: lib/processor/src/girouette/processor.clj [hunk 5]
```diff
@@ -157,7 +167,15 @@
         (pp/pprint @file-data file-writer))
       (let [all-css-classes (into #{} (mapcat :css-classes) (vals @file-data))
             predef-garden (if preflight? preflight [])
-            all-garden-defs (into predef-garden (keep (:garden-fn @config)) (sort all-css-classes))]
+            all-garden-defs (into predef-garden (keep (:garden-fn @config)) (sort all-css-classes))
+            all-garden-defs (reduce (fn [garden [cls classes]]
+                                      (into garden
+                                            (util/apply-class-rules
+                                              garden
```
Comment (green-coder) [inline lib/processor/src/girouette/processor.clj:174]: The process is in `n^2`, it will be too slow on larger sizes of inputs.

Comment (green-coder) [reply lib/processor/src/girouette/processor.clj:174]: I think I understand now the use of the `keep` in your code, it was to provide all the garden rules already processed by Girouette. I prefer to assume that all the garden rules passed to the apply-class-rules are used, and change the way the process is implemented.

File: lib/processor/src/girouette/processor.clj (hunk 3 above)
Comment (green-coder) [reply lib/processor/src/girouette/processor.clj:178]: `(requiring-resolve apply-classes)` throws an exception when apply-classes is `nil`. I will fix it, no problem.

**Pull request merged**